Assignment/FSMTest.c

```diff
@@ -0,0 +1,129 @@
+/// <param name="exeFilePathAndName"></param>
+/// <param name="inputFilePathAndName"></param>
+/// <param name="outputFilePathAndName"></param>
+/// <returns></returns>
+int ExecProgram (char* exeFilePathAndName, char* inputFilePathAndName, char* outputFilePathAndName) {
+   char* cmdline = malloc (strlen (exeFilePathAndName) + strlen (inputFilePathAndName) + strlen (outputFilePathAndName) + 3);
+   if (cmdline == NULL) {
+      printf ("Unable to allocate memory\n");
+      return 1;
+   }
+   sprintf (cmdline, "%s%s%s%s%s", exeFilePathAndName, " ", inputFilePathAndName, " ", outputFilePathAndName);
+
+   // Set up structures for process information
+   STARTUPINFOA si;
+   PROCESS_INFORMATION pi;
+
+   // Initialize memory for STARTUPINFO and PROCESS_INFORMATION structures
+   ZeroMemory (&si, sizeof (si));
+   si.cb = sizeof (si);
+   ZeroMemory (&pi, sizeof (pi));
+
+   // Attempt to create the process
+   if (!CreateProcessA (
+      NULL,            // Application name (NULL if using command line)
+      cmdline,          // Command line (path to the executable)
+      NULL,            // Process handle not inheritable
+      NULL,            // Thread handle not inheritable
+      FALSE,           // Set handle inheritance to FALSE
+      0,               // No creation flags
+      NULL,            // Use parent's environment block
+      NULL,            // Use parent's starting directory
+      &si,             // Pointer to STARTUPINFO structure
+      &pi))            // Pointer to PROCESS_INFORMATION structure
+   {
+      // If CreateProcess fails, print an error message
+      int err = GetLastError ();
+      printf ("Failed to start process. Error: %lu\n", GetLastError ());
+      free (cmdline);
+      return 1;
+   }
+
+   // Wait until the process has finished execution
+   WaitForSingleObject (pi.hProcess, INFINITE);
+
+   // Close handles
+   CloseHandle (pi.hProcess);
+   CloseHandle (pi.hThread);
+
+   free (cmdline);
+   return 0;
+
+}
+
+bool CompareFiles (char* file1, char* file2) {
+   FILE* f1 = fopen (file1, "r"), * f2 = fopen (file2, "r");
+   if (f1 == NULL || f2 == NULL) {
+      printf ("Error opening files\n");
+      return false;
+   }
+   int pos = 0, ch1, ch2;
+   while ((ch1 = fgetc (f1)) != EOF && (ch2 = fgetc (f2)) != EOF) {
+      if (ch1 != ch2) {
+         printf ("Error at bit no. %d, Expected %c, Actual %c\n", pos, ch1, ch2);
+         fclose (f1);
+         fclose (f2);
+         return false;
+      }
+      pos++;
+   }
+   fclose (f1);
+   fclose (f2);
+   return true;
+}
+
+/// <summary>
+/// Test Harness
+/// </summary>
+/// <param name="argc"></param>
+/// <param name="argv">argv[1] is the name of the FSM</param>
+/// <returns></returns>
+int main (int argc, char** argv) {
+#define NTESTS 7
+   if (argc != 2) {
+      printf ("Usage: %s <FSM executable name>\n", argv[0]);
+      return -1;
+   }
+   char* inputFiles[] = { "test1in.txt", "test2in.txt", "test3in.txt", "test4in.txt",

```

Comment:
Since the input file and reference file should not be modified, declare them as const char*, while you modify the output it can be declared as just char*